File: core/aarogya_core/types/ocr.py
"""OCR domain types."""

from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field

from aarogya_core.types.common import BoundingBox, SchemaVersionMixin


class PipelineStatus(StrEnum):
    """Pipeline execution status."""

    OK = "ok"
    NOT_IMPLEMENTED = "not_implemented"
    ERROR = "error"


class OCRWord(BaseModel):
    """A single recognized word."""

    text: str
    bbox: BoundingBox | None = None
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)


class OCRLine(BaseModel):
    """A line of recognized text."""

    text: str
    words: list[OCRWord] = Field(default_factory=list)
    bbox: BoundingBox | None = None
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)


class OCRRequest(SchemaVersionMixin):
    """Input for an OCR / HTR pipeline."""

    image_path: str | None = Field(default=None, description="Path relative to repo or absolute")
    image_base64: str | None = None
    engine_id: str | None = Field(default=None, description="e.g. paddle, qwen")
    config_ref: str | None = Field(default=None, description="Path under configs/")
    metadata: dict[str, Any] = Field(default_factory=dict)


class OCRResult(SchemaVersionMixin):
    """Output of an OCR / HTR pipeline."""

    status: PipelineStatus = PipelineStatus.NOT_IMPLEMENTED
    message: str | None = None
    full_text: str = ""
    lines: list[OCRLine] = Field(default_factory=list)
    engine_id: str | None = None
    config_hash: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)

File: shared/types/ocr.ts
import { z } from "zod";

import { SCHEMA_VERSION } from "../constants/api";
import { boundingBoxSchema } from "./common";

export const pipelineStatusSchema = z.enum(["ok", "not_implemented", "error"]);

export const ocrWordSchema = z.object({
  text: z.string(),
  bbox: boundingBoxSchema.nullable().optional(),
  confidence: z.number().min(0).max(1).nullable().optional(),
});

export const ocrLineSchema = z.object({
  text: z.string(),
  words: z.array(ocrWordSchema).default([]),
  bbox: boundingBoxSchema.nullable().optional(),
  confidence: z.number().min(0).max(1).nullable().optional(),
});

export const ocrRequestSchema = z.object({
  schema_version: z.literal(SCHEMA_VERSION).optional().default(SCHEMA_VERSION),
  image_path: z.string().nullable().optional(),
  image_base64: z.string().nullable().optional(),
  engine_id: z.string().nullable().optional(),
  config_ref: z.string().nullable().optional(),
  metadata: z.record(z.unknown()).optional().default({}),
});

export const ocrResultSchema = z.object({
  schema_version: z.literal(SCHEMA_VERSION).optional().default(SCHEMA_VERSION),
  status: pipelineStatusSchema,
  message: z.string().nullable().optional(),
  full_text: z.string().optional().default(""),
  lines: z.array(ocrLineSchema).optional().default([]),
  engine_id: z.string().nullable().optional(),
  config_hash: z.string().nullable().optional(),
  metadata: z.record(z.unknown()).optional().default({}),
});

export type OCRWord = z.infer<typeof ocrWordSchema>;
export type OCRLine = z.infer<typeof ocrLineSchema>;
export type OCRRequest = z.input<typeof ocrRequestSchema>;
export type OCRResult = z.infer<typeof ocrResultSchema>;
export type PipelineStatus = z.infer<typeof pipelineStatusSchema>;
